content/posts/csv-boilerplate-for-go.md

```diff
@@ -0,0 +1,307 @@
+
+Now we should think about testing. I'm a big fan of the [80/20 rule](https://en.wikipedia.org/wiki/Pareto_principle) when it comes to testing, so below is a test that covers the happy path very efficiently.
+
+```go
+func TestCSVReadWriteCycle(t *testing.T) {
+	in := strings.TrimSpace(`
+Car;MPG;Cylinders;Displacement;Horsepower;Weight;Acceleration;Model;Origin
+STRING;DOUBLE;INT;DOUBLE;DOUBLE;DOUBLE;DOUBLE;INT;CAT
+Chevrolet Chevelle Malibu;18.0;8;307.0;130.0;3504.;12.0;70;US
+Buick Skylark 320;15.0;8;350.0;165.0;3693.;11.5;70;US
+Plymouth Satellite;18.0;8;318.0;150.0;3436.;11.0;70;US
+`)
+	wantOut := strings.TrimSpace(`
+Car;MPG;Cylinders;Displacement;Horsepower;Weight;Acceleration;Model;Origin
+STRING;DOUBLE;INT;DOUBLE;DOUBLE;DOUBLE;DOUBLE;INT;CAT
+Chevrolet Chevelle Malibu;18.000000;8;307.000000;130.000000;3504.000000;12.000000;70;US
+Buick Skylark 320;15.000000;8;350.000000;165.000000;3693.000000;11.500000;70;US
+Plymouth Satellite;18.000000;8;318.000000;150.000000;3436.000000;11.000000;70;US
+`)
+
+	for i := 0; i < 2; i++ {
+		cars, err := ReadCarsCSV(strings.NewReader(in))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		buf := &bytes.Buffer{}
+		if err := WriteCarsCSV(buf, cars); err != nil {
+			t.Fatal(err)
+		}
+		gotOut := strings.TrimSpace(buf.String())
+		if gotOut != wantOut {
+			t.Fatalf("\ngot:\n%s\nwant:\n%s", gotOut, wantOut)
+		}
+		in = gotOut
+	}
+}
+```
+
+The test first reads the provided `in` CSV, and makes sure that converting it back to CSV results in `wantOut` which is a little different because of the way Go formats our floating point numbers. The test then proceeds to treat the output from the first iteration as the input for the second one. This makes sure that our implementation can read the original format, as well as the slightly different output it produces itself.
+
+Of course we could also aim to reproduce the float formatting quirks from the original file. At first glance it seems like all floats are formatted with one digit after the period. However, a closer look reveals that the `Weight` column has a period, but skips the following digit, e.g. `3504.`. I've decided to not go down this rabbit hole, but it should be clear that our boilerplate approach puts us in a great position for dealing with CSV quirks like this.
+
+Depending on how serious you are, you probably also want to test a few error cases and maybe even throw some fuzzing at this. But I've decided the test above is good enough for my boilerplate, so you'll have to do this part yourself.
+
+Anyway, thanks for taking the time to read this post. I make no claim that my approach for dealing with CSV is superior to all alternatives, but I think you could probably do a lot worse than starting with [my boilerplate](https://github.com/felixge/dump/blob/master/csv-boilerplate/main.go).
```

Comment:
Be bold :) No need for the disclaimer and the thank-you, even though it's friendly and kind.

Reply:
OK : ). 235da6bf488cc5e40fdb93b0242137856d636282